Fix caught objects disappearing when a replay is rewound

ekrctb opened the pull request:
Before this change, when a replay is rewound, caught objects were missing because caught objects are permanently removed after dropped from the catcher plate.
Also, objects that exploded from the plate were not properly removed and results in buggy floating objects.
Those issues are more noticeable in a catch editor (my WIP version). I thought solving this issue is important for the catch editor.

To manage caught objects efficiently, I used a `LifetimeEntryManager`. Each time an object is caught, a `CaughtObjectEntry` is created and added to `CaughtObjectContainer`. `CaughtObjectContainer` instantiates pooled `DrawableCaughtObject` from `CaughtObjectEntry` each time it became alive and remove when dead.
The logic is similar to the way `HitObjectContainer` uses `HitObjectLifetimeEntry`.

`CaughtObjectEntry` times number of hit objects are retained in memory for every gameplay, but the memory usage is not significant as I tested.

ekrctb committed:
Remove HitObject reference from CaughtObject

It is replaced by new CatchObjectType enum and StartTime.

ekrctb committed:
Introduce CaughtObjectContainer for lifetime management of caught objects

ekrctb committed:
Rename CaughtObject -> DrawableCaughtObject

ekrctb committed:
Fix caught object positioning

ekrctb committed:
Use stateless RNG for caught object stacking position computation

ekrctb committed:
Fix caught object not removed when rewound

ekrctb committed:
Add a test for caught object revert

ekrctb committed:
Remove debug printing

ekrctb committed:
Add test scene for CaughtObjectContainer

ekrctb committed:
Fix caught object has wrong colour

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/IHasCatchObjectState.cs:
```diff
@@ -12,7 +12,9 @@ namespace osu.Game.Rulesets.Catch.Objects.Drawables
     public interface IHasCatchObjectState
     {
-        PalpableCatchHitObject HitObject { get; }
+        CatchObjectType ObjectType { get; }
+
+        double StartTime { get; }
```
Immediately I'm not sure I agree with this premise of having that data added to all catch DHOs and copied to the caught object model. Why isn't this just the hitobject itself? Hitobjects (the models) are not pooled after all. Can't the dropped objects just take a reference to the HOs?

The main reason that I'm opposed to this is that catch object types are encoded already, in the hitobject types themselves. Adding this `ObjectType` enum to boot seems like unnecessary duplication - this sort of stuff should really have a singular source of ground truth.

I'm not sure if it's worth to review the rest of this since this is a pretty central piece of this entire change, so abstaining with the rest for now.

ekrctb replied:
The main motivation for removing `HitObject` reference was that its properties could change after the object is caught in the editor.
It turns out, however, in the current state of PR, `HitObject` properties being changed after caught doesn't cause issues. [The way](https://github.com/ppy/osu/blob/fa94703b9b663e443896a3971bfc2c2102350b93/osu.Game.Rulesets.Catch/Skinning/Default/CatchHitObjectPiece.cs#L51)`StartTime` currently used doesn't matter `StartTime` as long as it is after the current time. And the type of `HitObject` cannot be changed.
I'll make it references to a `HitObject` for now to make the PR simpler.

bdach replied:
A bit hard to me to respond to that without knowing exactly in which circumstances the properties could change. Let's just cross that bridge when we get to it, but I'd probably honestly rather prefer cloning hitobjects to this myself.

ekrctb committed:
Reference CatchHitObject from IHasCatchObjectState again.

The main motivation for removing HitObject reference was that its properties could change after the object is caught in the editor.
It turns out, however, in the current state of PR, HitObject properties being changed after caught doesn't cause issues. The way StartTime currently used doesn't matter StartTime as long as it is after the current time. And the type of HitObject cannot be changed.
I'll make it references to a HitObject for now to make the PR simpler.

ekrctb committed:
Remove debug printing

ekrctb committed:
Store drop position in CaughtObjectEntry

ekrctb committed:
Remove caught object when hit object is removed from playfield

ekrctb committed:
Store catcher direction when a caught object is dropped

ekrctb committed:
Fix wrong fruit type is used for caught objects

ekrctb committed:
Mirror dropped object to match mirroring of the catcher stack

ekrctb committed:
Add tests for dropped object position and mirroring

ekrctb committed:
Set default MirrorDirection

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+        private readonly DrawablePool<DrawableCaughtBanana> caughtBananaPool;
+        private readonly DrawablePool<DrawableCaughtDroplet> caughtDropletPool;
+
+        /// <summary>
+        /// The randomness used to compute position in stack.
+        /// It is incremented for each <see cref="Add"/> call and decremented for each <see cref="Remove"/> call to make a replay consistent.
+        /// </summary>
+        private int randomSeed = 1;
```
Might as well seed by index in beatmap at this point to save on tracking this.

ekrctb replied:
Because testing code doesn't set IndexInBeatmap, using it makes stacking ugly in testing.
It is fine to modify testing code, but I think tracking this value in `CaughtObjectContainer` (or maybe in `Catcher`) is better as it requires less preparation to use from outside.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+
+            if (entry.State != CaughtObjectState.Stacked) return;
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (entry.LifetimeStart <= Time.Current)
+                aliveStackedObjects.Add(entry);
```
This seems slightly suspicious and potentially buggy to me. Can whatever this is fixing be fixed in any other way, like (wild guess) scheduling stacked object drop after children, for instance?

ekrctb replied:
Multiple objects can be caught in one frame. `aliveStackObjects` affect stack positioning and we cannot just schedule the explosion to remove this code.
To update `aliveStackObjects` via `entryBecameAlive`, we need to call `lifetimeManager.Update` but this results in quadratic time complexity and is awkward (`LifetimeEntryManager` delays new entry to become alive for reasons).
My current stance is this way is the best.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+        {
+            randomSeed--;
+
+            lifetimeManager.RemoveEntry(entry);
+            removeDrawable(entry);
+
+            aliveStackedObjects.Remove(entry);
+
+            if (!dropEntryMap.TryGetValue(entry, out var dropEntry)) return;
+
+            dropEntryMap.Remove(entry);
+
+            lifetimeManager.RemoveEntry(dropEntry);
+            removeDrawable(dropEntry);
```
All of these fields are really hard to keep track of (what was added where and what needs to be cleaned up from where).

Would probably factor out four symmetrical methods, something like:

```csharp
public void addImmediateEntry(CaughtObjectEntry entry)
{
	lifetimeManager.AddEntry(entry);
	if (entry.State == CaughtObjectState.Stacked)
		addDelayedDropEntry(entry);
}

public void removeImmediateEntry(CaughtObjectEntry entry)
{
	lifetimeManager.RemoveEntry(entry);
	if (entry.State == CaughtObjectState.Stacked)
		removeDelayedDropEntry(entry);
}

public void addDelayedDropEntry(CaughtObjectEntry entry)
{
	var dropEntry = new CaughtObjectEntry(CaughtObjectState.Dropped, entry.PositionInStack, entry)
	{
		LifetimeStart = double.PositiveInfinity,
		LifetimeEnd = double.PositiveInfinity
	};

	lifetimeManager.AddEntry(dropEntry);
	dropEntryMap[entry] = dropEntry;
}

public void removeDelayedDropEntry(CaughtObjectEntry entry)
{
	dropEntryMap.Remove(entry, out var dropEntry);
	lifetimeManager.RemoveEntry(dropEntry);
}
```

I would hope that something can be done to eliminate the `aliveStackedObjects` adds/deletes here, such that that logic can be entirely located in `entryBecame{Alive,Dead}`.

Also not sure why `Remove()` is cleaning up drop entry drawables but `Add()` never adds them. If that is indeed required for some reason, then I'd leave a comment in `addDelayedDropEntry` as to why.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+            {
+                float diff = (displayRadius + radius_div_2) / allowance;
+
+                position.X += (StatelessRNG.NextSingle(randomSeed, iteration + 2) - 0.5f) * diff * 2;
```
guessing sequential values were intended here

```suggestion
                position.X += (StatelessRNG.NextSingle(randomSeed, iteration * 2) - 0.5f) * diff * 2;
```

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+            {
+                entry.LifetimeEnd = time;
+
+                if (!dropEntryMap.TryGetValue(entry, out var dropEntry)) continue;
```
This seems like it should be an assertion rather than a soft check. Can this realistically happen and not be a bug?

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,232 @@
+
+        private void removeDrawable(CaughtObjectEntry entry)
+        {
+            if (!drawableMap.TryGetValue(entry, out DrawableCaughtObject drawable))
+                return;
```
Can use `.Remove(entry, out var drawable)`, but again, I'd really be expecting this to never fail.

bdach commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -464,127 +469,62 @@ private void updateState(CatcherAnimationState state)
+                entry = new CaughtObjectEntry(CaughtObjectState.Dropped, positionInStack, hitObject)
+                {
+                    ApplyTransforms = applyExplodeTransforms,
+                    DropPosition = CaughtObjectContainer.GetCurrentDropPosition(positionInStack),
```
Seems weird for the catcher to have to call the container's own method like this. Should probably be `CaughtObjectContainer`'s responsibility to calculate this (which should hopefully allow `GetCurrentDropPosition` to become private).

ekrctb committed:
Merge branch 'master' into caught-object-container

ekrctb committed:
Move all caught object creation logic to CaughtObjectContainer

- The CaughtObjectEntry is now abstract and detailed logic of what need to be saved on this entry is entirely contained in the CaughtObjectContainer class.
  - Use StackedObjectEntry and DroppedObjectEntry to enforce type safety.
- Dropped object animation logic is also moved.
- HitObject->CaughtObjectEntry mapping feature is removed.
  - `Catcher.OnHitObjectRemoved` is also removed for now (it is not removing nested hit object anyway).

ekrctb committed:
Fix plus instead of times is used

ekrctb commented:
I moved all caught object creation logic to `CaughtObjectContainer`.
`CaughtObjectEntry` is now abstract and `CaughtObjectContainer` has `StackedObjectEntry` and `DroppedObjectEntry` private classes to store all internal data.
`CaughtObjectEntry` doesn't reference `DrawableCaughtObject` and it should apply transforms to a `Drawable` without knowing specifics of it being `DrawableCaughtObject`.

ekrctb commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectEntry.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public float DisplayRotation { get; }
+
+        public Bindable<FruitVisualRepresentation> VisualRepresentation { get; } = new Bindable<FruitVisualRepresentation>();
```
It is indeed possible to remove `VisualRepresentation` Bindable <https://github.com/ekrctb/osu/commit/e5ef1aefc0faeea03fcb4cb962042c559de4c834>.
However, it requires `IndexInBeatmap` bindable in `IHasCatchObjectState`.
Because `HitObject` referenced from `IHasObjectState` can change at anytime, a consumer cannot bind to `ObjectState.HitObject.SomeBindable` because it will become stale as soon as `ObjectState.HitObject` changes.

This is kind of bad design and I don't really like `IHasCatchObjectState` as it is very ad-hoc and there are still code duplication. Probably a better solution could be found in the future. But I don't touch it in this PR.
Also I made changes to made `CaughtObjectEntry` not implement `IHasCatchObjectState` so it is less relevant now.

bdach replied:
Yeah, my original comment was more of a passing remark, not something to be addressed in this PR.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/IHasCatchObjectState.cs:
```diff
@@ -12,7 +12,7 @@ namespace osu.Game.Rulesets.Catch.Objects.Drawables
     public interface IHasCatchObjectState
     {
-        PalpableCatchHitObject HitObject { get; }
+        CatchHitObject HitObject { get; }
```
Not sure where this generalisation is coming from - I think this should remain as `Palpable...`? This surely doesn't apply to banana showers or juice streams, right?

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/DrawableCaughtObject.cs:
```diff
@@ -41,23 +34,26 @@ protected CaughtObject(CatchSkinComponents skinComponent, Func<ISkinComponent, D
+            AccentColour.Value = entry.AccentColour;
+            HyperDash.Value = entry.HyperDash;
+
+            Scale = Vector2.Divide(entry.DisplaySize, Size) * 0.5f;
```
Where is this `0.5f` coming from? I'm not sure I saw it be there before.

ekrctb replied:
It was `caughtObject.Scale /= 2` <https://github.com/ppy/osu/pull/11255/files#diff-c3ab09080a53b7f21d7ceffeaf2e5bfd4ab7ce5c5262696ea8d725197f87c640L476>.

bdach commented on osu.Game.Rulesets.Catch/Judgements/CatchJudgementResult.cs:
```diff
@@ -20,6 +21,8 @@ public class CatchJudgementResult : JudgementResult
         /// </summary>
         public bool CatcherHyperDash;
 
+        public CaughtObjectEntry CaughtObjectEntry;
```
Not 100% happy with the layering/separation of concerns here... Should the judgement result really be aware of when/how it was caught via the entry? Maybe this would be better suited to the HO itself? Or be tracked somewhere external entirely, as I think was the case (dictionaries in `CaughtObjectContainer`)?

ekrctb replied:
A caught object is a result of a judgment (catching) so maintaining the caught object entry makes sense to me.
I cannot be a property of the HO because a hit object can be associated with multiple judgment results (not currently supported in `Playfield`, but used in tests, and I think it makes sense conceptually).
A caught object could be associated with `HitObjectLifetimeEntry` but there is no code path to modify this object before #11262.

bdach replied:
Do you have a preview or other loose branch of how this would look with that mentioned pull? That might provide better rationale for having that merged.

ekrctb replied:
I wrote a prototype of associating `CaughtObjectEntry` with `HitObjectLifetimeEntry` <https://github.com/ekrctb/osu/commit/392d048fe2e73236262427dc5bf0667b9b0ad31a>.
I think this way is not great. It need to expose `LifetimeEntry` in DHO to make it accessible.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,334 @@
+            }
+        }
+
+        private class StackedObjectEntry : CaughtObjectEntry
```
Would split these nested classes/enums to separate files just to keep line count of this file down.

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,334 @@
+            addImmediateEntry(stackEntry);
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (stackEntry.LifetimeStart <= Clock.CurrentTime)
```
I don't know if it's just me but I'd prefer this conditional was flipped, as it feels more natural for me to have the instant being checked on the left side of the comparison against a start time of an interval (was part of the reason why I misread earlier and thought this was buggy).

bdach commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs:
```diff
@@ -0,0 +1,334 @@
+            return position;
+        }
+
+        public void DropStackedObjects(DroppedObjectAnimation animation, int mirrorDirection)
```
nit: would unify up naming here (there's `AddStackObject` and `AddDropObject`, but `DropStackedObject`).

bdach commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -464,127 +461,29 @@ private void updateState(CatcherAnimationState state)
-            caughtObject.Anchor = Anchor.TopCentre;
-            caughtObject.Position = position;
-            caughtObject.Scale /= 2;
+            var entry = source.HitObject is Droplet
+                ? CaughtObjectContainer.AddDropObject(source, positionInStack, DroppedObjectAnimation.Explode, stackMirrorDirection)
+                : CaughtObjectContainer.AddStackObject(source, positionInStack);
```
This logic should probably be in `CaughtObjectContainer` (it's even duplicated in tests). Seems like there can possibly be only one method that accepts a caught HO, maybe.

ekrctb committed:
Revert unnecessary change of IHasCatchObjectState

ekrctb committed:
Remove redundant check of adding stacked object to the alive set.

ekrctb committed:
Normalize names and add doc comments to the public methods of CaughtObjectContainer.

ekrctb committed:
Move "stack or immediate drop for droplet" logic to `CaughtObjectContainer`.

ekrctb committed:
Split CaughtObjectContainer nested classes to separate files.

peppy committed:
Merge branch 'master' into caught-object-container

ekrctb commented:
It will be rewritten.
